### PR TITLE
Fix Weird Metrics

### PR DESCRIPTION
**Pull request by kmada97**
Initial fixes to metrics that were not matching prometheus standards.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -35,7 +35,7 @@ vet:
 	@echo ">> vetting code"
 	@$(GO) vet $(pkgs)
 
-megacheck:
+megacheck:
```
Comment (joehandzik) [inline Makefile]: Any particular reason for this space?

File: sources/pmwebapi.go [hunk 2]
```diff
@@ -110,8 +110,29 @@ func typeLabel(units string, pcpType string, name string) string {
...
+	name = strings.Replace(name, "failcnt", "failcount", -1)
+	name = strings.Replace(name, "Mbyte", "megabyte", -1)
+	name = strings.Replace(name, "byte", "bytes", -1)
+	name = strings.Replace(name, "sec", "seconds", -1)
+	name = strings.Replace(name, "bytes_byte", "bytes", -1)
+	name = strings.Replace(name, "bytess_bytes", "bytes", -1)
+	name = strings.Replace(name, "megabytesss", "megabytes", -1)
+	name = strings.Replace(name, "kilobytesss", "kilobytes", -1)
+	name = strings.Replace(name, "millisecondsonds", "milliseconds", -1)
+	name = strings.Replace(name, "nanosecondsonds", "nanoseconds", -1)
```
Comment (joehandzik) [inline sources/pmwebapi.go]: Aren't these supposed to be "nanosecondonds" instead of "nanosecondsonds" (and so on)? I still feel like there's a smarter way to do this than making multiple fixes on the same metric to fix what the first fix breaks for some metrics. Any good ideas here, @roclark ?

File: sources/pmwebapi_test.go [hunk 3]
```diff
@@ -92,12 +161,12 @@ func TestLabelTypes(t *testing.T) {
...
 
-	units = "87"
-	pcpType = "countER"
-	name = "k.a.l"
-	response = typeLabel(units, pcpType, name)
-	expected = "k_a_l_87_total"
-	if response != expected {
-		t.Fatalf("Failed when there were periods in the name. Expected: %s, Got: %s", expected, response)
-	}
+	//units = "87"
```
Comment (joehandzik) [inline sources/pmwebapi_test.go]: Probably wanna delete these if you don't want them anymore rather than commenting them out.

File: sources/pmwebapi_test.go [hunk 4]
```diff
@@ -54,7 +54,76 @@ func TestUnmarshal(t *testing.T) {
 		t.Fatalf("Not receiving correct URL. Expected: %s, Got: %s", expected, testMap.URL)
 	}
 }
+func TestWeirdMetrics(t *testing.T) {
```
Comment (joehandzik) [inline sources/pmwebapi_test.go]: You could create slices or arrays with both the name and the expected value and iterate through using a loop. Would save code space and minimize redundancy at the same time.

File: sources/pmwebapi.go (hunk 2 above)
Comment (roclark) [reply sources/pmwebapi.go]: There's a bigger issue here. Look at the following example:
Say we have a metric name which is simple: `milliseconds` - there is nothing wrong with this name. Going through your chain, it will hit line 121 and replace `millisec` with `milliseconds`, giving you `millisecondsonds`. Then, line 125 will match again because it would replace `sec` with `seconds`, technically giving you `millisecondsondsonds`. Finally, line 130 updates this to `millisecondsonds` for your final answer. So... this actually still doesn't work properly. You can verify by adding another unit test case where you test `milliseconds` and it will throw an error saying that it retrieved `millisecondsonds`.

Long story short, if you end up using this method, you will need to think critically about the order of operations. Since `sec` is in the middle, and `sec` is in multiple names down below, it will keep updating itself. One way is to work on the lowest common denominators first (such as `sec` and `byte`) and then update the other ones later.

Also, remember to look at the worst case scenario for your unit tests. Yes, testing the specific names seen here should always work, but don't forget to test things that _aren't_ listed here that might not work. As I mentioned, even `milliseconds` will give issues.

File: sources/pmwebapi_test.go [hunk 5]
```diff
@@ -54,7 +54,76 @@ func TestUnmarshal(t *testing.T) {
...
+	if response != expected {
+		t.Fatalf("Failed when name was byte. Expected: %s, Got: %s", expected, response)
+	}
+	name = "megabytess"
+	response = fixNaming(name)
+	expected = "megabytes"
+	if response != expected {
+		t.Fatalf("Failed when name was Mbyte. Expected: %s, Got: %s", expected, response)
+	}
+
```
Comment (roclark) [inline sources/pmwebapi_test.go]: Do you need this space?

File: sources/pmwebapi_test.go (hunk 4 above)
Comment (roclark) [inline sources/pmwebapi_test.go]: This could probably be much cleaner and more efficient. 90% of the code here is repeated multiple times, and the remaining 10% can easily be set as variables. I would recommend creating an array of the tested and expected results then iterating through them. This function could be about 10 lines long, while still easily adding new scenarios without copying yourself.

Comment (roclark) [reply sources/pmwebapi_test.go]: Also, as mentioned before, remember to test _other_ scenarios as well - ones that would potentially break this. That's what the point of unit tests is about really. Remember, you always expect a new car to work in perfect weather, but how does it handle and perform when it's snowing, raining, freezing, or in sweltering heat?

File: sources/pmwebapi.go (hunk 2 above)
Comment (roclark) [reply sources/pmwebapi.go]: I will keep thinking on this and see if there is a more efficient way to catch these, but at the very least, what I mentioned so far is a good start.

**Review by roclark — CHANGES_REQUESTED** [review]
Probably going to take a few more updates, but this is more along the lines of what we are looking for.

File: sources/pmwebapi_test.go (hunk 4 above)
Comment (roclark) [reply sources/pmwebapi_test.go]: Also also, add a newline before this function - currently it runs into the previous one.

File: sources/pmwebapi_test.go [hunk 6]
```diff
@@ -54,7 +54,76 @@ func TestUnmarshal(t *testing.T) {
 		t.Fatalf("Not receiving correct URL. Expected: %s, Got: %s", expected, testMap.URL)
 	}
 }
+func TestWeirdMetrics(t *testing.T) {
+	name := "count / sec"
+	response := fixNaming(name)
+	expected := "count_per_second"
+	if response != expected {
+		t.Fatalf("Failed when name was count / second. Expected: %s, Got: %s", expected, response)
+	}
```
Comment (roclark) [inline sources/pmwebapi_test.go]: Do you need this space?

File: sources/pmwebapi.go [hunk 7]
```diff
@@ -110,8 +110,29 @@ func typeLabel(units string, pcpType string, name string) string {
...
+
+	name = strings.Replace(name, "count_per_secondsond", "count_per_second", -1)
 	return strings.Replace(name, ".", "_", -1)
+
```
Comment (roclark) [inline sources/pmwebapi.go]: Do you need this space?

File: Makefile (hunk 1 above)
Comment (kmada97) [reply Makefile]: I think that was an accident, I'll change it

File: sources/pmwebapi_test.go (hunk 4 above)
Comment (kmada97) [reply sources/pmwebapi_test.go]: That sounds like a better idea, ill do that

File: sources/pmwebapi_test.go [hunk 8]
```diff
@@ -55,49 +55,29 @@ func TestUnmarshal(t *testing.T) {
...
-	units = "tery"
-	pcpType = "gagage"
-	name = "kal"
-	response = typeLabel(units, pcpType, name)
-	expected = "kal_tery"
-	if response != expected {
-		t.Fatalf("Failed when type is any other than counter or empty. Expected: %s, Got: %s", expected, response)
+	name := []string{"count / sec", "Kbyte", "nanosec", "millisec", "failcnt", "byte", "megabytess", "Mbyte", "sec", "bytes_byte", "kilobytess"}
+	expected := []string{"count_per_second", "kilobytes", "nanoseconds", "milliseconds", "failcount", "bytes", "megabytes", "megabytes", "seconds", "bytes", "kilobytes"}
+	var n = 0
```
Comment (roclark) [inline sources/pmwebapi_test.go]: Remember, this isn't C... High-level languages have come a long way since C and we can now use handy tools like `range`. Go makes this even easier because you can get both the iteration counter and the array item in a single line/call.

File: sources/pmwebapi_test.go [hunk 9]
```diff
@@ -55,49 +55,29 @@ func TestUnmarshal(t *testing.T) {
...
+	for n < len(name) {
+		if fixNaming(name[n]) != expected[n] {
+			t.Fatalf("Failed when name was %s. Expected: %s, Got: %s", name[n], expected[n], fixNaming(name[n]))
```
Comment (roclark) [inline sources/pmwebapi_test.go]: It would probably be best to set the result of `fixNaming(name[n])` to a variable instead of calling it twice. Say the `fixNaming` function throws an error, this line could do something very odd. Not necessarily critical, but not a bad practice.

File: sources/pmwebapi_test.go [hunk 10]
```diff
@@ -55,49 +55,29 @@ func TestUnmarshal(t *testing.T) {
...
-	expected = "k_a_l_87_total"
-	if response != expected {
-		t.Fatalf("Failed when there were periods in the name. Expected: %s, Got: %s", expected, response)
+func TestLabelTypes(t *testing.T) {
+	units := []string{"", "", "kal", "tery"}
+	pcpType := []string{"", "", "CoUnTeR", "gagage"}
+	name := []string{"kal", "", "kal", "kal"}
+	expected := []string{"kal", "", "kal_kal_total", "kal_tery"}
+	var n = 0
+	for n < len(name) {
```
Comment (roclark) [inline sources/pmwebapi_test.go]: Again, this isn't C. Use the `range` tool or something else available in Go.

File: sources/pmwebapi_test.go [hunk 11]
```diff
@@ -55,49 +55,29 @@ func TestUnmarshal(t *testing.T) {
...
-	}
-
-	units = "tery"
-	pcpType = "gagage"
-	name = "kal"
-	response = typeLabel(units, pcpType, name)
-	expected = "kal_tery"
-	if response != expected {
-		t.Fatalf("Failed when type is any other than counter or empty. Expected: %s, Got: %s", expected, response)
+	name := []string{"count / sec", "Kbyte", "nanosec", "millisec", "failcnt", "byte", "megabytess", "Mbyte", "sec", "bytes_byte", "kilobytess"}
```
Comment (roclark) [inline sources/pmwebapi_test.go]: This is one way to do it... but it might be easier and cleaner to have a map or an array of structs (though probably the former). That way, you can clearly see which names align with which expected results. Right now, it's not easy to quickly identify which variables map together in these arrays. Also, if these ever got out of order, it would be painful.

File: sources/pmwebapi.go [hunk 12]
```diff
@@ -110,7 +110,38 @@ func typeLabel(units string, pcpType string, name string) string {
...
+	name = strings.Replace(name, "byte", "bytes", -1)
+	name = strings.Replace(name, "sec", "seconds", -1)
+	name = strings.Replace(name, "bytes_byte", "bytes", -1)
+	name = strings.Replace(name, "bytess_bytes", "bytes", -1)
+	name = strings.Replace(name, "megabytesss", "megabytes", -1)
+	name = strings.Replace(name, "kilobytesss", "kilobytes", -1)
+	name = strings.Replace(name, "millisecondsonds", "milliseconds", -1)
+	name = strings.Replace(name, "nanosecondsonds", "nanoseconds", -1)
+	name = strings.Replace(name, "count_per_secondsond", "count_per_second", -1)
+	name = strings.Replace(name, "BLOCK", "block", -1)
```
Comment (roclark) [inline sources/pmwebapi.go]: Technically speaking, this and the following lines work... buuuuut you can save yourself a lot of space, make this look cleaner, and catch potentially missed metric names by just using something from the strings package like `ToLower()`.

**Review by roclark — CHANGES_REQUESTED** [review]
Still a good amount more to do here, but we are getting closer. Although, I still agree with @joehandzik's original suggestion for using a `case` or `if...else` block to catch the names instead of your current method. That will make life a lot easier as this might actually be breaking more metric names than it is fixing.

**Comment by joehandzik** [comment]
@roclark @kmada97 I agree with all of Robert's latest comments and don't have anything more to add.

File: sources/pmwebapi_test.go [hunk 13]
```diff
@@ -55,49 +55,28 @@ func TestUnmarshal(t *testing.T) {
...
-	expected = "kal_kal_total"
-	if response != expected {
-		t.Fatalf("Failed when type was case sensitive. Expected: %s, Got: %s", expected, response)
+	name := []string{"count / sec", "Kbyte", "nanosec", "millisec", "failcnt", "byte", "megabytess", "Mbyte", "sec", "bytes_byte", "kilobytess"}
+	expected := []string{"count_per_second", "kilobytes", "nanoseconds", "milliseconds", "failcount", "bytes", "megabytes", "megabytes", "seconds", "bytes", "kilobytes"}
+	for n, _ := range name {
+		if fixNaming(name[n]) != expected[n] {
+			t.Fatalf("Failed when name was %s. Expected: %s, Got: %s", name[n], expected[n], fixNaming(name[n]))
+		}
+		n++
```
Comment (roclark) [inline sources/pmwebapi_test.go]: Is this necessary? If you tried to access `name[n]`, or `expected[n]` after this line you will get a panic.

File: sources/pmwebapi_test.go [hunk 14]
```diff
@@ -55,49 +55,28 @@ func TestUnmarshal(t *testing.T) {
...
-	name = "k.a.l"
-	response = typeLabel(units, pcpType, name)
-	expected = "k_a_l_87_total"
-	if response != expected {
-		t.Fatalf("Failed when there were periods in the name. Expected: %s, Got: %s", expected, response)
+	for n, _ := range name {
+		if typeLabel(units[n], pcpType[n], name[n]) != expected[n] {
+			t.Fatalf("Failed when units was %s, Type was %s, and name was %s. Expected: %s, Got: %s", units[n], pcpType[n], name[n], expected[n], typeLabel(units[n], pcpType[n], name[n]))
+		}
+		n++
```
Comment (roclark) [inline sources/pmwebapi_test.go]: See the other comment about removing `n++`.

File: sources/pmwebapi_test.go [hunk 15]
```diff
@@ -55,49 +55,28 @@ func TestUnmarshal(t *testing.T) {
...
-	units = "kal"
-	pcpType = "CoUnTeR"
-	name = "kal"
-	response = typeLabel(units, pcpType, name)
-	expected = "kal_kal_total"
-	if response != expected {
-		t.Fatalf("Failed when type was case sensitive. Expected: %s, Got: %s", expected, response)
+	name := []string{"count / sec", "Kbyte", "nanosec", "millisec", "failcnt", "byte", "megabytess", "Mbyte", "sec", "bytes_byte", "kilobytess"}
+	expected := []string{"count_per_second", "kilobytes", "nanoseconds", "milliseconds", "failcount", "bytes", "megabytes", "megabytes", "seconds", "bytes", "kilobytes"}
+	for n, _ := range name {
```
Comment (roclark) [inline sources/pmwebapi_test.go]: When you convert the arrays to a map, you will no longer iterate over `n, _` and instead something like `_, result`. This is still currently a C implementation and not true golang semantics.

File: sources/pmwebapi_test.go [hunk 16]
```diff
@@ -55,49 +55,28 @@ func TestUnmarshal(t *testing.T) {
...
+	expected := []string{"kal", "", "kal_kal_total", "kal_tery"}
 
-	units = "87"
-	pcpType = "countER"
-	name = "k.a.l"
-	response = typeLabel(units, pcpType, name)
-	expected = "k_a_l_87_total"
-	if response != expected {
-		t.Fatalf("Failed when there were periods in the name. Expected: %s, Got: %s", expected, response)
+	for n, _ := range name {
```
Comment (roclark) [inline sources/pmwebapi_test.go]: See other comment about what to iterate over.

File: sources/pmwebapi.go [hunk 17]
```diff
@@ -110,8 +110,27 @@ func typeLabel(units string, pcpType string, name string) string {
...
+	name = strings.Replace(name, "sec", "seconds", -1)
+	name = strings.Replace(name, "bytes_byte", "bytes", -1)
+	name = strings.Replace(name, "bytess_bytes", "bytes", -1)
+	name = strings.Replace(name, "megabytesss", "megabytes", -1)
+	name = strings.Replace(name, "kilobytesss", "kilobytes", -1)
+	name = strings.Replace(name, "millisecondsonds", "milliseconds", -1)
+	name = strings.Replace(name, "nanosecondsonds", "nanoseconds", -1)
+	name = strings.Replace(name, "count_per_secondsond", "count_per_second", -1)
+	name = strings.Replace(name, ".", "_", -1)
+	return strings.ToLower(name)
```
Comment (roclark) [inline sources/pmwebapi.go]: This `ToLower(name)` call should be the first thing you do in this function.

File: sources/pmwebapi_test.go [hunk 18]
```diff
@@ -55,49 +55,28 @@ func TestUnmarshal(t *testing.T) {
...
+func TestWeirdMetrics(t *testing.T) {
 
-	units = "kal"
-	pcpType = "CoUnTeR"
-	name = "kal"
-	response = typeLabel(units, pcpType, name)
-	expected = "kal_kal_total"
-	if response != expected {
-		t.Fatalf("Failed when type was case sensitive. Expected: %s, Got: %s", expected, response)
+	name := []string{"count / sec", "Kbyte", "nanosec", "millisec", "failcnt", "byte", "megabytess", "Mbyte", "sec", "bytes_byte", "kilobytess"}
```
Comment (roclark) [inline sources/pmwebapi_test.go]: When you add the switchcases, we are going to need to see more testcases here. Check for names that are already good and make sure they aren't broken. This should be a fairly long list.

File: sources/pmwebapi.go [hunk 19]
```diff
@@ -110,8 +110,27 @@ func typeLabel(units string, pcpType string, name string) string {
 	} else {
 		name += unit
 	}
+	return name
+}
+
+func fixNaming(name string) string {
 	name = strings.Replace(name, "count / sec", "count_per_second", -1)
-	return strings.Replace(name, ".", "_", -1)
+	name = strings.Replace(name, "Kbyte", "kilobyte", -1)
```
Comment (roclark) [inline sources/pmwebapi.go]: Note that when you convert these to switchcases, you will no longer require every single one of these items. Some of the later ones are a result of these earlier actions. I will let you figure out which ones you need to keep.

**Review by roclark — CHANGES_REQUESTED** [review]
Looking forward to seeing the switchcases and how that changes things up for you.

File: sources/pmwebapi.go [hunk 20]
```diff
@@ -110,7 +110,44 @@ func typeLabel(units string, pcpType string, name string) string {
...
 		name += unit
 	}
-	name = strings.Replace(name, "count / sec", "count_per_second", -1)
+	return name
+}
+
+func fixNaming(name string) string {
+	name = strings.ToLower(name)
+	switch name {
+	case "seconds":
```
Comment (roclark) [inline sources/pmwebapi.go]: Do you need any of these blank ones? This just adds more complexity to the switchcase and more steps for the code to step through. I won't mark all of these, but I don't think we need them.

File: sources/pmwebapi.go [hunk 21]
```diff
@@ -110,7 +110,44 @@ func typeLabel(units string, pcpType string, name string) string {
...
+	case "megabytes":
+
+	case "bytes":
+
+	case "bytes_byte":
+		name = strings.Replace(name, "bytes_byte", "bytes", -1)
+	case "byte":
+		name = strings.Replace(name, "byte", "bytes", -1)
+	default:
+		name = strings.Replace(name, "failcnt", "failcount", -1)
```
Comment (roclark) [inline sources/pmwebapi.go]: Why is your default case to replace `failcnt` with `failcount`. I like to interpret the default case as the situation that doesn't follow special parsing or rules. So, the `failcnt` should probably a case statement of its own, and you probably don't even need a default in this scenario because all other instances stay the same.

File: sources/pmwebapi_test.go [hunk 22]
```diff
@@ -55,49 +55,41 @@ func TestUnmarshal(t *testing.T) {
...
+	name["Mbyte"] = "megabytes"
+	name["sec"] = "seconds"
+	name["bytes_byte"] = "bytes"
+	name["kilobytes"] = "kilobytes"
+	name["seconds"] = "seconds"
+	name["nanoseconds"] = "nanoseconds"
+	name["milliseconds"] = "milliseconds"
+	name["megabytes"] = "megabytes"
+	name["bytes"] = "bytes"
+	for k := range name {
```
Comment (roclark) [inline sources/pmwebapi_test.go]: This looks much better as a map than the previous array. But, to make this even more go-centric, when you iterate over a map with `range`, you iterate over the `key` first, then the `value` second. In other words, if you need both the key and value (which in your case you do), then it is best to setup your for loop as (don't use the variable names `key` and `value` like I do here, use something more appropriate for your setup):
```golang
for key, value := range m {
    ...
}
```
This way, you can replace `name[k]` with `value` (again, using a different naming scheme) for direct access.

File: sources/pmwebapi_test.go [hunk 23]
```diff
@@ -55,49 +55,41 @@ func TestUnmarshal(t *testing.T) {
...
+		if fixNaming(k) != name[k] {
+			t.Fatalf("Failed when name was %s. Expected: %s, Got: %s", k, name[k], fixNaming(k))
```
Comment (roclark) [inline sources/pmwebapi_test.go]: Remember to set the result of `fixNaming(k)` to a variable before the `if` statement, and compare that instead of making the call in this function again. This will prevent any weirdness with a return or errors.

File: sources/pmwebapi.go [hunk 24]
```diff
@@ -110,7 +110,44 @@ func typeLabel(units string, pcpType string, name string) string {
 	} else {
 		name += unit
 	}
-	name = strings.Replace(name, "count / sec", "count_per_second", -1)
+	return name
+}
+
+func fixNaming(name string) string {
+	name = strings.ToLower(name)
+	switch name {
```
Comment (roclark) [inline sources/pmwebapi.go]: This is forward progress, but I don't think this actually does what you expect it to do. Try testing with something like `disk_access_time_nanosec`. The result _should_ be `disk_access_time_nanoseconds`, but the way this is written, it will still be `disk_access_time_nanosec`. I will leave this as an exercise for you to figure out what's wrong, but let me know if you get absolutely stuck.

File: sources/pmwebapi_test.go [hunk 25]
```diff
@@ -55,49 +55,41 @@ func TestUnmarshal(t *testing.T) {
...
+	name := make(map[string]string)
 
-	units = "kal"
-	pcpType = "CoUnTeR"
-	name = "kal"
-	response = typeLabel(units, pcpType, name)
-	expected = "kal_kal_total"
-	if response != expected {
-		t.Fatalf("Failed when type was case sensitive. Expected: %s, Got: %s", expected, response)
+	name["count / sec"] = "count_per_second"
```
Comment (roclark) [inline sources/pmwebapi_test.go]: Good job including the metric names that are valid to make sure they work as expected, but look at my previous message (on why the switchcase won't parse as expected) to think of longer names. Though it's great to check the name "seconds", you likely won't have "seconds" as an actual PCP metric name. Try using longer ones too to make sure this actually works as expected.

File: sources/pmwebapi_test.go (hunk 22 above)
Comment (roclark) [inline sources/pmwebapi_test.go]: Also, I personally would add a newline before this loop.

File: sources/pmwebapi_test.go [hunk 26]
```diff
@@ -55,49 +55,41 @@ func TestUnmarshal(t *testing.T) {
...
-	units = ""
-	pcpType = ""
-	name = ""
-	response = typeLabel(units, pcpType, name)
-	expected = ""
-	if response != expected {
-		t.Fatalf("Failed when all strings were empty. Expected: %s, Got: %s", expected, response)
-	}
+func TestWeirdMetrics(t *testing.T) {
+	name := make(map[string]string)
```
Comment (roclark) [inline sources/pmwebapi_test.go]: This is a styling preference, but I personally would remove this newline.

**Review by roclark — CHANGES_REQUESTED** [review]
Forward progress on this one, but still a little bit more to do. Let me know if you have questions on any of this, but it should be enough to get you moving further.

**Comment by roclark** [comment]
Also, it doesn't look like your commits are linked to your GitHub account. Are you using your GitHub account on your server for the commits?

File: sources/pmwebapi.go (hunk 20 above)
Comment (joehandzik) [inline sources/pmwebapi.go]: This is the right idea. Do you have the case statements for cases that require no update because without them you catch their substrings still? If so, I'd say at least remove the newlines that you have placed below cases that require no changes. It's worth taking a look at this link (https://github.com/golang/go/wiki/Switch) for other places you can make optimizations. For example, you could collapse "bytes_byte" and "byte" into a single case statement since they require the same change.

I'm also wondering if this code actually works...you're passing the entire name as the switch case, which I don't think will have an exact match to any of your cases. You're likely going to need to search for the case substrings you've identified within the larger string, I figure.

Comment (kmada97) [reply sources/pmwebapi.go]: I added these blank ones, because the way I understood it, the switch case will exit if it finds the correct case, so if seconds is written correctly in a metric I dont want it to hit the case for sec and get changed again so putting this makes the cases that are already right exit.

File: sources/pmwebapi.go (hunk 21 above)
Comment (kmada97) [reply sources/pmwebapi.go]: I got it, went ahead and changed this

File: sources/pmwebapi_test.go (hunk 22 above)
Comment (kmada97) [reply sources/pmwebapi_test.go]: for this would i defrine value as value = name[k] and use that?

Comment (roclark) [reply sources/pmwebapi_test.go]: No need to define anything, the `for key, value := range m` example I gave you automatically gives you the key (analogous to your `k`) and the value (analogous to your `name[k]`) for every item of your map. These two are always paired as well, so you don't have to worry about getting a value that doesn't match the key.

File: sources/pmwebapi.go (hunk 20 above)
Comment (kmada97) [reply sources/pmwebapi.go]: the cases that require no update are the ones I left blank
and youre right I need to find a way to search for the smaller substring in the larger string

Comment (roclark) [reply sources/pmwebapi.go]: Switch statements are very similar to `if..else` statements, so once it finds a case that matches, it exits the block. The only difference is when you use [fallthrough](https://github.com/golang/go/wiki/Switch#fall-through) which can still match the other cases. In your case, you do not want fallthrough (and you don't have it implemented anyway), so you can expect that if it matches one of your cases, it won't evaluate any of the others.

File: sources/pmwebapi_test.go [hunk 27]
```diff
@@ -55,49 +62,69 @@ func TestUnmarshal(t *testing.T) {
...
 
-	units = "kal"
-	pcpType = "CoUnTeR"
-	name = "kal"
-	response = typeLabel(units, pcpType, name)
-	expected = "kal_kal_total"
-	if response != expected {
-		t.Fatalf("Failed when type was case sensitive. Expected: %s, Got: %s", expected, response)
-	}
+var tester []testlables
```
Comment (roclark) [inline sources/pmwebapi_test.go]: Global variable???

File: sources/pmwebapi_test.go [hunk 28]
```diff
@@ -55,49 +62,69 @@ func TestUnmarshal(t *testing.T) {
...
 	}
-
-	units = "87"
-	pcpType = "countER"
-	name = "k.a.l"
-	response = typeLabel(units, pcpType, name)
-	expected = "k_a_l_87_total"
-	if response != expected {
-		t.Fatalf("Failed when there were periods in the name. Expected: %s, Got: %s", expected, response)
+	for _ := range tester {
```
Comment (roclark) [inline sources/pmwebapi_test.go]: I'm assuming you are going to fix this too?

File: sources/pmwebapi_test.go [hunk 29]
```diff
@@ -55,49 +62,69 @@ func TestUnmarshal(t *testing.T) {
...
 
-	units = "tery"
-	pcpType = "gagage"
-	name = "kal"
-	response = typeLabel(units, pcpType, name)
-	expected = "kal_tery"
-	if response != expected {
-		t.Fatalf("Failed when type is any other than counter or empty. Expected: %s, Got: %s", expected, response)
+func TestLabelTypes(t *testing.T) {
+	tester = []testlables{
```
Comment (roclark) [inline sources/pmwebapi_test.go]: Good to use an array of structs, but you can just declare the struct at the top of the file, and then do the initializations in one line. Something like...
```golang
type structName struct {
    name string
    value int
    length int
}

...

arr := []structName{
    {"first", 1, 10},
    {"second", 7, 3},
}
```

File: sources/pmwebapi_test.go (hunk 28 above)
Comment (kmada97) [reply sources/pmwebapi_test.go]: yup

File: sources/pmwebapi_test.go [hunk 30]
```diff
@@ -23,6 +23,13 @@ type AutoGenerated struct {
 	URL    string `json:"url"`
 }
 
+type testlables struct {
+	name     string
+	tyype    string
```
Comment (roclark) [inline sources/pmwebapi_test.go]: Why is this `tyype`?

File: sources/pmwebapi_test.go [hunk 31]
```diff
@@ -55,49 +62,51 @@ func TestUnmarshal(t *testing.T) {
...
 	}
 
-	units = "87"
-	pcpType = "countER"
-	name = "k.a.l"
-	response = typeLabel(units, pcpType, name)
-	expected = "k_a_l_87_total"
-	if response != expected {
-		t.Fatalf("Failed when there were periods in the name. Expected: %s, Got: %s", expected, response)
+	for range arr {
```
Comment (roclark) [inline sources/pmwebapi_test.go]: I'm pretty confused as to what's going on with this function. Do you mind explaining? I don't think you're actually iterating values like you might expect...

**Review by roclark — CHANGES_REQUESTED** [review]
At the moment, just another function and a variable name to cleanup.

We also want to think about the switchcase more - at present, it works for single instances, but think of how something like `read_rate_KByte_per_sec`. How would that one end up? That could be saved for a follow-up PR though.

File: sources/pmwebapi_test.go (hunk 30 above)
Comment (kmada97) [reply sources/pmwebapi_test.go]: when i wrote type it gave me an error, type is specific in golang i think

File: sources/pmwebapi_test.go (hunk 31 above)
Comment (kmada97) [reply sources/pmwebapi_test.go]: im going through the range of the array with all thevalues 
at least thats what im trying to do

**Comment by roclark** [comment]
Remove the `pcp_exporter` binary and the second commit ("Merge branch 'master' into weirdmetrics") and you should be good.

**Comment by joehandzik** [comment]
@kmada97 You still have three total commits. Please merge them to we can merge your PR!

**Comment by joehandzik** [comment]
@kmada97 This rebase has changes in it that you didn't write in this pull request, which probably means that you rebased in older commits. You are likely going to need to do a diff between your working branch and master, then create a new branch with an accurate single commit. I'll see if there's a smarter solution, but for now assume that that's the best you've got.

**Comment by roclark** [comment]
@joehandzik it could just be that he needs to rebase against master again. If you look at the [network](https://github.com/HewlettPackard/pcp_exporter/network), it shows the `weirdmetrics` branch is based off a reeeaaaallly old commit (ie. one of the first ones).

**Comment by joehandzik** [comment]
@roclark Maybe. It's worth a shot! Would be faster than what I suggested.

**Comment by joehandzik** [comment]
@kmada97 Your CI build is still failing. You need to do a test build before checking in to figure out what is wrong. You should avoid checking in before doing a test build, ever.

File: sources/pmwebapi_test.go [hunk 32]
```diff
@@ -23,6 +23,13 @@ type AutoGenerated struct {
 	URL    string `json:"url"`
 }
 
+type testlables struct {
+	Units    string `json:"units"`
```
Comment (roclark) [inline sources/pmwebapi_test.go]: Technically, the `testlables` struct isn't JSON, but just a string, so it would probably be best to get rid of the `json:"units"` part. Though it's not a huge deal.

File: sources/pmwebapi_test.go [hunk 33]
```diff
@@ -23,6 +23,13 @@ type AutoGenerated struct {
 	URL    string `json:"url"`
 }
 
+type testlables struct {
```
Comment (roclark) [inline sources/pmwebapi_test.go]: This should be `testLabels` instead to use go-semantics and proper spelling.

**Review by roclark — COMMENTED** [review]
It'd be nice to see the `testlables` name changed, but otherwise this looks good to me!

**Review by roclark — APPROVED** [review]

**Comment by roclark** [comment]
LGTM now! @joehandzik?
